src/web-extensions.c

```diff
@@ -0,0 +1,794 @@
+
+  if (session)
+    g_debug ("webextensions session owned by '%s' created", session->sender);
+
+  web_extensions_session = (WebExtensionsSession *)session;
+  web_extensions_session->mode = mode;
+  return web_extensions_session;
+}
+
+static gboolean
+handle_create_session (XdpDbusWebExtensions *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(GError) error = NULL;
+  Session *session;
+
+  session = (Session *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_host_exited (GPid pid,
+                gint status,
+                gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  web_extensions_session->child_pid = -1;
+  web_extensions_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  /* This regexp comes from the Mozilla documentation on valid native
+     messaging host names:
+
+     https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#native_messaging_manifests
+
+     That is, one or more dot-separated groups composed of
+     alphanumeric characters and underscores.
+  */
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (WebExtensionsSessionMode mode)
+{
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  search_path = g_ptr_array_new_with_free_func (g_free);
+  switch (mode)
+    {
+    case WEB_EXTENSIONS_SESSION_MODE_CHROMIUM:
+      /* Chrome and Chromium search paths documented here:
+       * https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-location
```

Comment:
```suggestion
       * https://developer.chrome.com/docs/extensions/nativeMessaging/#native-messaging-host-location
```

Fixed documentation link. "apps" is deprecated in Chrome and should not be referenced.

Reply:
Done.